core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -989,17 +725,501 @@ private static SortOrder freshSortOrder(int orderId, Schema schema, SortOrder so
+        setCurrentSnapshot(null, System.currentTimeMillis());
+      }
+
+      return this;
+    }
+
+    public Builder setProperties(Map<String, String> updated) {
+      if (updated.isEmpty()) {
+        return this;
+      }
+
+      properties.putAll(updated);
+      changes.add(new MetadataUpdate.SetProperties(updated));
+
+      return this;
+    }
+
+    public Builder removeProperties(Set<String> removed) {
+      if (removed.isEmpty()) {
+        return this;
+      }
+
+      removed.forEach(properties::remove);
+      changes.add(new MetadataUpdate.RemoveProperties(removed));
+
+      return this;
+    }
+
+    public Builder setLocation(String newLocation) {
+      if (location != null && !location.equals(newLocation)) {
+        return this;
       }
+
+      this.location = newLocation;
+      changes.add(new MetadataUpdate.SetLocation(newLocation));
+
+      return this;
+    }
+
+    public TableMetadata build() {
+      if (changes.size() == startingChangeCount) {
+        return base;
+      }
+
+      if (lastUpdatedMillis == null) {
+        this.lastUpdatedMillis = System.currentTimeMillis();
+      }
+
+      Schema schema = schemasById.get(currentSchemaId);
+      PartitionSpec.checkCompatibility(specsById.get(defaultSpecId), schema);
+      SortOrder.checkCompatibility(sortOrdersById.get(defaultSortOrderId), schema);
+
+      List<MetadataLogEntry> metadataHistory = addPreviousFile(previousFileLocation, base.lastUpdatedMillis());
+
+      // find intermediate snapshots to suppress incorrect entries in the snapshot log.
+      //
+      // transactions can create snapshots that are never the current snapshot because several changes are combined
+      // by the transaction into one table metadata update. when each intermediate snapshot is added to table metadata,
+      // it is added to the snapshot log, assuming that it will be the current snapshot. when there are multiple
+      // snapshot updates, the log must be corrected by suppressing the intermediate snapshot entries.
+      //
+      // a snapshot is an intermediate snapshot if it was added but is not the current snapshot.
+      Set<Long> addedSnapshotIds = Sets.newHashSet();
+      Set<Long> intermediateSnapshotIds = Sets.newHashSet();
+      for (MetadataUpdate update : changes) {
+        if (update instanceof MetadataUpdate.AddSnapshot) {
+          // adds must always come before set current snapshot
+          MetadataUpdate.AddSnapshot addSnapshot = (MetadataUpdate.AddSnapshot) update;
+          addedSnapshotIds.add(addSnapshot.snapshot().snapshotId());
+        } else if (update instanceof MetadataUpdate.SetCurrentSnapshot) {
+          long snapshotId = ((MetadataUpdate.SetCurrentSnapshot) update).snapshotId();
+          if (addedSnapshotIds.contains(snapshotId) && snapshotId != currentSnapshotId) {
+            intermediateSnapshotIds.add(snapshotId);
+          }
+        }
+      }
+
+      // update the snapshot log
```

Comment:
The only update to this block is checking whether a snapshot was intermediate and suppressing the log for it.